homeassistant/components/crownstone/entry_manager.py

```diff
@@ -0,0 +1,226 @@
+
+        # create listener for when home assistant is stopped
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.on_shutdown)
+
+        # register all entities
+        self.hass.config_entries.async_setup_platforms(self.config_entry, PLATFORMS)
+
+        return True
+
+    async def async_process_events(self, sse_client: CrownstoneSSEAsync) -> None:
+        """Asynchronous iteration of Crownstone SSE events."""
+        async with sse_client as client:
+            async for event in client:
+                if event is not None:
+                    # make SSE updates, like ability change, available to the user
+                    self.hass.bus.async_fire(f"{DOMAIN}_{event.type}", event.data)
+
+    async def async_unload(self) -> bool:
+        """Unload the current config entry."""
+        # authentication failed
+        if self.cloud.cloud_data is None:
+            return True
+
+        # close sse client and unsub from listeners
+        self.sse.close_client()
+        for sse_unsub in self.listeners[SSE_LISTENERS]:
+            sse_unsub()
+
+        # close uart connection and unsub from listeners
+        if hasattr(self, ATTR_UART):
+            self.uart.stop()
+            for subscription_id in self.listeners[UART_LISTENERS]:
+                UartEventBus.unsubscribe(subscription_id)
+
+        # unload all platform entities
+        unload_ok = await self.hass.config_entries.async_unload_platforms(
+            self.config_entry, PLATFORMS
+        )
+
+        if unload_ok:
+            self.hass.data[DOMAIN].pop(self.config_entry.entry_id)
+
+        return unload_ok
+
+    @callback
+    def on_shutdown(self, _: Event) -> None:
+        """Close all IO connections."""
+        self.sse.close_client()
+        if hasattr(self, ATTR_UART):
+            self.uart.stop()
+
+
+async def async_update_entry_options(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Handle options update."""
+    entry_data = entry.data.copy()
+    # USB was configured at setup & user wants to remove config
+    if (
+        not entry.options.get(CONF_USE_CROWNSTONE_USB)
+        and entry.data.get(CONF_USB_PATH) is not None
+    ):
+        entry_data[CONF_USB_PATH] = None
+
+        # update and reload
+        # USB connection is closed and entry will be created without USB connection
+        hass.config_entries.async_update_entry(entry, data=entry_data)
+        await hass.config_entries.async_reload(entry.entry_id)
+
+    # USB was not configured at setup & user wants to configure
+    # Init config flow on USB configuration step
+    elif (
+        entry.options.get(CONF_USE_CROWNSTONE_USB)
+        and entry.data.get(CONF_USB_PATH) is None
+    ):
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": "usb_config"},
+                data={CONF_UNIQUE_ID: entry.entry_id},
+            )
+        )
```

Comment:
Don't start a new flow here. The removal and/or addition should happen as additional the options flow. 


```suggestion
```

Reply:
I assume the same step (code) that is used in the config flow for the usb currently, but living in the options flow? Do you know an integration that does this that i could look at as an example?

Reply:
I couldn't find an example that lined up with this use case particularly well, but the basic idea is to break out as much of the functionality that isn't specific to driving the flow itself into help functions so they can be reused between the config flow and the options flow